src/timeline/files.rs

```diff
@@ -0,0 +1,90 @@
+use crate::timeline::FFIZER_DATASTORE_DIRNAME;
+use crate::Result;
+
+use std::collections::HashMap;
+use std::fs;
+use std::path::Path;
+
+use cid::Cid;
+use multihash_codetable::{Code, MultihashDigest};
+
+const FILES_FILENAME: &str = "files.json";
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub(crate) struct FileMeta {
+    pub key: String,
+    pub hash: Cid,
+}
+
+impl FileMeta {
+    pub fn with_key(self, key: &String) -> Self {
+        Self {
+            key: key.to_owned(),
+            ..self
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+struct TrackedFiles {
+    pub files: HashMap<String, HashMap<String, (FileMeta, FileMeta)>>,
+}
+
+pub(crate) fn save_metas_for_source<T>(infos: T, target_folder: &Path, source: String) -> Result<()>
+where
+    T: IntoIterator<Item = (FileMeta, FileMeta)>,
+{
+    let mut tracked = load_tracked(target_folder)?;
+    let mut map: HashMap<String, (FileMeta, FileMeta)> = HashMap::new();
```

Comment:
🎯 suggestion: 
If you don't want to do the check, but are sure the keys are identical, use an `assert_eq!(key0, key1);` (or `debug_assert_eq!` if you want to skip the check in release mode)

Otherwise you can write

```rust
    let mut map = HashMap::new();
    for (info0, info1) in infos {
        if info0.key != info1.key {
            continue; // maybe return an error?
        }
        map.insert(info0.key.clone(), (info0, info1));
    }
```

An alternative could be 

```rust
let map = infos
        .into_iter()
        .filter_map(|(info0, info1)| {
            if info0.key == info1.key {
                Some((info0.key.clone(), (info0, info1)))
            } else {
                None
            }
        })
        .collect();
```